<a href="https://colab.research.google.com/github/manjulamishra/DS-Unit-1-Sprint-4-Statistical-Tests-and-Experiments/blob/master/Manjula_DS_Unit_1_Sprint_Challenge_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 4

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 (the class attribute) is the most interesting, as it separates the 307 approved cases from the 383 rejected cases. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [0]:
#load the dataset
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data'
import pandas as pd
df = pd.read_csv(url, header = None)


In [7]:
df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

In [0]:
#adding column names
column_names = ['A1',
'A2-continuous',
'A3-continuous',
'A4',
'A5',
'A6',
'A7',
'A8-continuous',
'A9',
'A10',
'A11-continuous',
'A12',
'A13',
'A14-continuous',
'A15-continuous',
'approved-not_approved']

In [0]:
#checking if column names are added
df = pd.read_csv(url, header = None, names = column_names)

In [47]:
df.head()

A1 A2-continuous  A3-continuous A4 A5 A6 A7  A8-continuous A9 A10  \
0  b         30.83          0.000  u  g  w  v           1.25  t   t   
1  a         58.67          4.460  u  g  q  h           3.04  t   t   
2  a         24.50          0.500  u  g  q  h           1.50  t   f   
3  b         27.83          1.540  u  g  w  v           3.75  t   t   
4  b         20.17          5.625  u  g  w  v           1.71  t   f   

   A11-continuous A12 A13 A14-continuous  A15-continuous approved-not_approved  
0               1   f   g          00202               0                     +  
1               6   f   g          00043             560                     +  
2               0   f   g          00280             824                     +  
3               5   t   g          00100               3                     +  
4               0   f   s          00120               0                     +

In [13]:
#checking the shape if it matches the shape described in the raw file
df.shape #it matches

(690, 16)

In [48]:
#to check NaN 
df.isna().sum() #it doesn't show because they might not be coded nans

A1                       0
A2-continuous            0
A3-continuous            0
A4                       0
A5                       0
A6                       0
A7                       0
A8-continuous            0
A9                       0
A10                      0
A11-continuous           0
A12                      0
A13                      0
A14-continuous           0
A15-continuous           0
approved-not_approved    0
dtype: int64

In [0]:
# taking care of the missing values
df = pd.read_csv(url, header = None, names = column_names, na_values=["?"], usecols=range(0,16))

In [50]:
df.isna().sum()

A1                       12
A2-continuous            12
A3-continuous             0
A4                        6
A5                        6
A6                        9
A7                        9
A8-continuous             0
A9                        0
A10                       0
A11-continuous            0
A12                       0
A13                       0
A14-continuous           13
A15-continuous            0
approved-not_approved     0
dtype: int64

In [24]:
df.dtypes

A1                        object
A2-continuous            float64
A3-continuous            float64
A4                        object
A5                        object
A6                        object
A7                        object
A8-continuous            float64
A9                        object
A10                       object
A11                        int64
A12                       object
A13                       object
A14-continuous           float64
A15-continuous             int64
approved-not_approved     object
dtype: object

In [0]:
df = df.dropna(0)

In [78]:
df.isna().sum()

A1                       0
A2-continuous            0
A3-continuous            0
A4                       0
A5                       0
A6                       0
A7                       0
A8-continuous            0
A9                       0
A10                      0
A11-continuous           0
A12                      0
A13                      0
A14-continuous           0
A15-continuous           0
approved-not_approved    0
dtype: int64

In [0]:
df['A15-continuous'] = df['A15-continuous'].astype(float)


In [0]:
df['A11-continuous'] = df['A11-continuous'].astype(float)

In [53]:
df.dtypes

A1                        object
A2-continuous            float64
A3-continuous            float64
A4                        object
A5                        object
A6                        object
A7                        object
A8-continuous            float64
A9                        object
A10                       object
A11-continuous           float64
A12                       object
A13                       object
A14-continuous           float64
A15-continuous             int64
approved-not_approved     object
dtype: object

In [80]:
import numpy as np
df['approved-not_approved'].value_counts()

-    357
+    296
Name: approved-not_approved, dtype: int64

### Raw File Describe Vs Our Investigation
File size: 690 
Class Distribution (arrpoved or not approved)
  
    +: 307 (44.5%)
    -: 383 (55.5%)
   
   A1:  12
    A2:  12
    A4:   6
    A5:   6
    A6:   9
    A7:   9
   A14: 13
   
### From our priliminary investigation, we have the right number of observations and missing values, also we changed column 15 from integer to float to make it a continuous variable 

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" between them and the class label, and apply the Chi-squared test to them. There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end. This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [81]:
df.describe()

A2-continuous  A3-continuous  A8-continuous  A11-continuous  \
count     653.000000     653.000000     653.000000      653.000000   
mean       31.503813       4.829533       2.244296        2.502297   
std        11.838267       5.027077       3.371120        4.968497   
min        13.750000       0.000000       0.000000        0.000000   
25%        22.580000       1.040000       0.165000        0.000000   
50%        28.420000       2.835000       1.000000        0.000000   
75%        38.250000       7.500000       2.625000        3.000000   
max        76.750000      28.000000      28.500000       67.000000   

       A14-continuous  A15-continuous  
count      653.000000      653.000000  
mean       180.359877     1013.761103  
std        168.296811     5253.278504  
min          0.000000        0.000000  
25%         73.000000        0.000000  
50%        160.000000        5.000000  
75%        272.000000      400.000000  
max       2000.000000   100000.000000

In [82]:
import numpy as np
df['approved-not_approved'].value_counts()

-    357
+    296
Name: approved-not_approved, dtype: int64

In [83]:
df['A1'].value_counts()

b    450
a    203
Name: A1, dtype: int64

In [116]:
df_table =  pd.crosstab(df['A1'], df['approved-not_approved'])
df_table

approved-not_approved    +    -
A1                             
a                       95  108
b                      201  249

In [118]:
pd.crosstab(df['A4'], df['approved-not_approved'])

approved-not_approved    +    -
A4                             
l                        2    0
u                      249  250
y                       45  107

In [0]:
df_plus = df.loc[df['approved-not_approved'] == '+']

In [87]:
df_plus.head()

A1  A2-continuous  A3-continuous A4 A5 A6 A7  A8-continuous A9 A10  \
0  b          30.83          0.000  u  g  w  v           1.25  t   t   
1  a          58.67          4.460  u  g  q  h           3.04  t   t   
2  a          24.50          0.500  u  g  q  h           1.50  t   f   
3  b          27.83          1.540  u  g  w  v           3.75  t   t   
4  b          20.17          5.625  u  g  w  v           1.71  t   f   

   A11-continuous A12 A13  A14-continuous  A15-continuous  \
0             1.0   f   g           202.0             0.0   
1             6.0   f   g            43.0           560.0   
2             0.0   f   g           280.0           824.0   
3             5.0   t   g           100.0             3.0   
4             0.0   f   s           120.0             0.0   

  approved-not_approved  
0                     +  
1                     +  
2                     +  
3                     +  
4                     +

In [88]:
df_plus.shape

(296, 16)

In [107]:
df.mean()

A2-continuous       31.503813
A3-continuous        4.829533
A8-continuous        2.244296
A11-continuous       2.502297
A14-continuous     180.359877
A15-continuous    1013.761103
dtype: float64

In [108]:
df_plus.mean()

A2-continuous       33.845473
A3-continuous        5.971943
A8-continuous        3.475186
A11-continuous       4.716216
A14-continuous     164.621622
A15-continuous    2009.726351
dtype: float64

In [109]:
df_negative.mean()

A2-continuous      29.562269
A3-continuous       3.882325
A8-continuous       1.223725
A11-continuous      0.666667
A14-continuous    193.408964
A15-continuous    187.974790
dtype: float64

In [0]:
df_negative = df.loc[df['approved-not_approved'] == '-']

In [95]:
df_negative.head()

A1  A2-continuous  A3-continuous A4 A5  A6  A7  A8-continuous A9 A10  \
70  b          32.33           7.50  u  g   e  bb          1.585  t   f   
72  a          38.58           5.00  u  g  cc   v         13.500  t   f   
73  b          44.25           0.50  u  g   m   v         10.750  t   f   
74  b          44.83           7.00  y  p   c   v          1.625  f   f   
75  b          20.67           5.29  u  g   q   v          0.375  t   t   

    A11-continuous A12 A13  A14-continuous  A15-continuous  \
70             0.0   t   s           420.0             0.0   
72             0.0   t   g           980.0             0.0   
73             0.0   f   s           400.0             0.0   
74             0.0   f   g           160.0             2.0   
75             1.0   f   g           160.0             0.0   

   approved-not_approved  
70                     -  
72                     -  
73                     -  
74                     -  
75                     -

In [89]:
df_negative.shape

(383, 16)

In [0]:
df.

In [96]:
df_negative.isna().sum()

A1                       0
A2-continuous            0
A3-continuous            0
A4                       0
A5                       0
A6                       0
A7                       0
A8-continuous            0
A9                       0
A10                      0
A11-continuous           0
A12                      0
A13                      0
A14-continuous           0
A15-continuous           0
approved-not_approved    0
dtype: int64

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is extremely different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

In [90]:
import scipy
df_plus_ttest = scipy.stats.ttest_1samp(df_plus.iloc[:,range(1,2)], 0.5)
print(df_plus_ttest)

Ttest_1sampResult(statistic=array([45.21089738]), pvalue=array([1.15298297e-134]))


In [97]:
import scipy
df_negative_ttest = scipy.stats.ttest_1samp(df_negative.iloc[:,range(1,2)], )
print(df_negative_ttest)

Ttest_1sampResult(statistic=array([51.22742676]), pvalue=array([2.4803241e-166]))


###Two different columns A2 and A3 from the same file

In [105]:
scipy.stats.ttest_ind(df_plus['A2-continuous'], df_plus['A3-continuous'])

Ttest_indResult(statistic=34.68218079542805, pvalue=1.6152941893381163e-144)

###Two same columns A2 from different datasets

In [106]:
scipy.stats.ttest_ind(df_plus['A2-continuous'], df_negative['A2-continuous'])

Ttest_indResult(statistic=4.675662433767456, pvalue=3.5636710933835225e-06)

##Two same columns A15 from different datasets (they both have drastically signiifcant mean)

In [110]:
scipy.stats.ttest_ind(df_plus['A15-continuous'], df_negative['A15-continuous'])

Ttest_indResult(statistic=4.475369764700449, pvalue=9.003915641872878e-06)

###Two same columns A14 from different datasets (they both have simliar means)

In [111]:
scipy.stats.ttest_ind(df_plus['A14-continuous'], df_negative['A14-continuous'])

Ttest_indResult(statistic=-2.18221936438708, pvalue=0.029450100409286473)

For the categorical features, explore by creating "cross tabs" between them and the class label, and apply the Chi-squared test to them. There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

In [0]:
from scipy.stats import chisquare

In [136]:
A4 = pd.crosstab(df['A4'], df['approved-not_approved'])
print(chisquare(A4, axis=None))

Power_divergenceResult(statistic=614.800918836141, pvalue=1.2814886836851252e-130)


In [137]:
A1 = pd.crosstab(df['A1'], df['approved-not_approved'])
print(chisquare(A1, axis=None))

Power_divergenceResult(statistic=101.00306278713629, pvalue=9.458202945413513e-22)


In [139]:
A5 = pd.crosstab(df['A5'], df['approved-not_approved'])
print(chisquare(A5, axis=None))

Power_divergenceResult(statistic=614.800918836141, pvalue=1.2814886836851252e-130)


In [140]:
A6 = pd.crosstab(df['A6'], df['approved-not_approved'])
print(chisquare(A6, axis=None))

Power_divergenceResult(statistic=376.9969372128637, pvalue=2.3658930783071553e-63)


In [141]:
A7 = pd.crosstab(df['A7'], df['approved-not_approved'])
print(chisquare(A7, axis=None))

Power_divergenceResult(statistic=1768.9509954058194, pvalue=0.0)


In [142]:
A9 = pd.crosstab(df['A9'], df['approved-not_approved'])
print(chisquare(A9, axis=None))

Power_divergenceResult(statistic=354.32006125574276, pvalue=1.730751761550774e-76)


In [143]:
A10 = pd.crosstab(df['A10'], df['approved-not_approved'])
print(chisquare(A10, axis=None))

Power_divergenceResult(statistic=152.16385911179174, pvalue=8.994150430808612e-33)


In [144]:
A12 = pd.crosstab(df['A12'], df['approved-not_approved'])
print(chisquare(A12, axis=None))

Power_divergenceResult(statistic=11.47166921898928, pvalue=0.009430579350519394)


In [147]:
A13 = pd.crosstab(df['A13'], df['approved-not_approved'])
print(chisquare(A13, axis=None))

Power_divergenceResult(statistic=1011.8851454823891, pvalue=1.605961312555435e-216)


## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

###- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?

I selected the two sets of two columns: one wooked similar and the other which drastically different in two data sets.
First, I splitted the data in to two different sets:
1: with all + data
2: all - (negative) data

I assume that postive(+) means that the people were approved their credit card application and negative means they weren't approved. 

I, then, calculated the mean of the main datafram before the split and of the spiltted dataframes. Based, on the individual column mean, I picked the columns which had almost the similar means (A2, A14) in both splitted df and the columns which had dramatically different means (A15) 

#### Results of A2:
Ttest_indResult(statistic=4.675662433767456, pvalue=3.5636710933835225e-06)
P-value is signlificantly low and thus we reject the null hypothesis . For example: if A2 represented income of people from both data sets, out null hypothesis was that income doesn't affect the credit card approval decision. In this case, we reject the null and say that income indeed was a decisive factor

#### Results of A15:
Ttest_indResult(statistic=4.475369764700449, pvalue=9.003915641872878e-06)

This column had a very different mean in two data sets
here again, our p value is very very low so we 